tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/legalize_to_linalg.cc

```diff
@@ -912,6 +912,96 @@ class TransposeConverter
+      ConversionPatternRewriter& rewriter) const final {
+    Location loc = real_dynamic_slice_op.getLoc();
+    auto arg_type = adaptor.operand().getType().dyn_cast<ShapedType>();
+    if (!arg_type || !arg_type.hasRank()) {
+      return rewriter.notifyMatchFailure(real_dynamic_slice_op,
+                                         "require known-rank args");
+    }
+    auto result_type =
+        this->typeConverter->convertType(real_dynamic_slice_op.getType())
+            .cast<RankedTensorType>();
+    Type index_type = rewriter.getIndexType();
+    Value zero =
+        rewriter.create<arith::ConstantOp>(loc, rewriter.getI64IntegerAttr(0));
+    Type i64_type = rewriter.getI64Type();
+    SmallVector<OpFoldResult, 4> offsets, sizes, strides;
+    SmallVector<Type, 3> clamp_type(3, i64_type);
+    for (auto i : llvm::seq<unsigned>(0, arg_type.getRank())) {
+      Value dim =
+          rewriter.create<arith::ConstantOp>(loc, rewriter.getIndexAttr(i));
+      Value start =
+          rewriter.create<tensor::ExtractOp>(loc, adaptor.start_indices(), dim);
+      Value limit =
+          rewriter.create<tensor::ExtractOp>(loc, adaptor.limit_indices(), dim);
+      Value stride =
+          rewriter.create<tensor::ExtractOp>(loc, adaptor.strides(), dim);
+
+      // Compute i-th dimension size of the result : size[i].
+      // If the i-th dimension of the result type is known, we go ahead with it
+      // else we compute it using limit, start and stride values.
+      int64_t result_dim_size = result_type.getDimSize(i);
+      Value size;
+      if (result_dim_size != -1)
+        size = rewriter.create<arith::ConstantOp>(
+            loc, rewriter.getIndexAttr(result_dim_size));
+      else
+        size = computeSize(loc, start, limit, stride, rewriter);
+
+      // Fetch i-th dimension size of the operand and calculate upper bound as
+      // :-
+      //    ub = operand_dim[i] - size[i]
+      Value operand_dim_size =
+          rewriter.createOrFold<tensor::DimOp>(loc, adaptor.operand(), dim);
+      Value upper_bound =
+          rewriter.createOrFold<arith::SubIOp>(loc, operand_dim_size, size);
+
+      // We clamp the start_index to keep it bounded as :-
+      // start index : 0 <= start_index[i] <= ub
+      // ClampOp lowering does not support index type, so we cast it into
+      // integer type.
+      start = rewriter.create<arith::IndexCastOp>(loc, i64_type, start);
+      upper_bound =
+          rewriter.create<arith::IndexCastOp>(loc, i64_type, upper_bound);
+      start = mhlo::MhloOpToStdScalarOp::map<mhlo::ClampOp>(
+          loc, i64_type, clamp_type, ValueRange{zero, start, upper_bound},
+          &rewriter);
+
+      offsets.push_back(
+          rewriter.create<arith::IndexCastOp>(loc, index_type, start)
+              .getResult());
+      if (result_dim_size != -1)
+        sizes.push_back(rewriter.getI64IntegerAttr(result_dim_size));
+      else
+        sizes.push_back(size);
```

Comment:
Isn't the type of `size` index type? In this case, we have to cast it to I64 type, right?

I think we don't need an if statement here. Instead, we can create a CastOp, and it should work for both cases.

I believe that canonicalizer would fold the cast op away if the input is a constant op.

Reply:
Yes, `size` is index type.
But only for those dimensions where the slice size is statically known do we require `I64Attr` and not `I64` type. For others we require them to indeed be in `index` type.

```
Tensor::ExtractSliceOp :- 

let arguments = (ins
    AnyRankedTensor:$source,
    Variadic<Index>:$offsets,
    Variadic<Index>:$sizes,
    Variadic<Index>:$strides,
    I64ArrayAttr:$static_offsets,
    I64ArrayAttr:$static_sizes,
    I64ArrayAttr:$static_strides
  );
 ```
 The build function I'm using is : `rewriter.replaceOpWithNewOp<tensor::ExtractSliceOp>(
        real_dynamic_slice_op, result_type, adaptor.operand(), offsets, sizes,
        strides);`

I believe we'd require the `if` to have the statically known slice sizes as `I64Attr`, something we can't afford to exercise for dynamically computed slice sizes.

Reply:
I see, then we don't need the cast op because index type is required. In this case, we don't really need `I64Attr` because canonicalizer can handle this case.

It's fine to do it this way, but I'd suggest not. Because this makes logic/implementation more complicated.

https://github.com/llvm/llvm-project/blob/1e09f0a97271352435b3858ca9d2ae47f491760a/mlir/test/Dialect/Tensor/canonicalize.mlir#L336-L351

Also please wrap if-else statements with braces.

```cpp
if (...) {
 ...
} else {
 ...
}

Reply:
The resulting type of  `size` via `result_dim_size` or `ceil((limit - start)/stride)` is `index` type indeed :-
```
Size[0] => %c256 = arith.constant 256 : index
Size[1] => %12 = arith.ceildivui %11, %10 : index
```

So, if we are to use `index` type itself then we can get rid of the `if-else` clause totally and simply write `sizes.push_back(size)`.

And I did try that initially but it threw a result size inference mismatch error :- 
```
error: expected type to be 'tensor<?x?xf32>' or a rank-reduced version. (size mismatch)

%26 = "tensor.extract_slice"(%arg0, %12, %25, %5, %18, %4, %16) {
                              operand_segment_sizes = dense<[1, 2, 2, 2]> : vector<4xi32>,
                              static_offsets = [-9223372036854775808, -9223372036854775808],
                              static_sizes = [-1, -1],
                              static_strides = [-9223372036854775808, -9223372036854775808]
             } : (tensor<256x?xf32>, index, index, index, index, index, index) -> tensor<256x?xf32>
```

If I use `I64Attr` then the `static_sizes` is found to be `[256, -1]` and the inference turns out to be successful.

Can you tell me what am I missing here?

Reply:
Oops, it might be the issue in passing result type. There are known values in result type, but nothing in static_operands. To make it work, I think we'll have to pass a whole dynamic type (i.e., tensor<?x?x?xf32>). This will need another if-else, and did not make things simple. So I think we're good at current patch. Sorry for that bad suggestion.